File: site/cgi/post.py
import base64
import os
import sys
import json

print("Content-Type: text/html\r\n\r\n")

print("<html>")
print("<head>")
print("<title>Hello - CGI POST</title>")
print("</head>")
print("<body>")
print("<h2>Hello CGI POST</h2>")

print("<h3>Environment Variables:</h3>")
print("<pre>")
for var_name, var_value in os.environ.items():
    print(f"{var_name} = {var_value}")
print("</pre>")

# Get form data from POST request
try:
    content_length = int(os.environ.get('CONTENT_LENGTH', 0))
except ValueError:
    content_length = 0

if content_length > 0:
    post_data = sys.stdin.read(content_length)
    form_data = json.loads(post_data)
    
    # Print form data
    print("<h3>POST Data:</h3>")
    print("<pre>")
    for key, value in form_data.items():
        print(f"{key} = {value}")
    print("</pre>")

    if "name" in form_data:
        name = form_data["name"]
        print("<h3>Hello {}!</h3>".format(name))
    else:
        print("<h3>Error: name not provided!</h3>")
else:
    print("<h3>Error: no data received!</h3>")

print("</body>")
print("</html>")

File: site/cgi/post.js
const url = require('url');

// Read POST data from stdin if the request method is POST
let postData = '';
if (process.env.REQUEST_METHOD === 'POST') {
  process.stdin.setEncoding('utf8');
  process.stdin.on('readable', () => {
    const chunk = process.stdin.read();
    if (chunk !== null) {
      postData += chunk;
    }
  });
}

process.stdin.on('end', () => {
  // Parse the POST data into an object
  let data = {};
  if (postData) {
    try {
      data = JSON.parse(postData);
    } catch (error) {
      console.error('Invalid JSON data in POST request');
    }
  }

  // Generate the HTML response
  let response = '<html>';
  response += '<head>';
  response += '<title>Hello - CGI Node.js Script</title>';
  response += '</head>';
  response += '<body>';
  response += '<h2>Hello CGI</h2>';

  if (Object.keys(data).length > 0) {
    response += '<h3>Data:</h3>';
    response += '<pre>';
    for (let key in data) {
      response += `${key} = ${data[key]}<br>`;
    }
    response += '</pre>';

    if (data.name) {
      response += `<h3>Hello ${data.name}!</h3>`;
    } else {
      response += '<h3>btw what\'s your name ?!</h3>';
    }
  } else {
    response += '<h3>No data received!</h3>';
  }

  response += '</body>';
  response += '</html>';

  // Send the response to stdout
  console.log('Content-Type: text/html\r\n\r\n');
  console.log(response);
});

